jetty-websocket/javax-websocket-server/src/main/java/org/eclipse/jetty/websocket/javax/server/JavaxWebSocketServerContainer.java

```diff
@@ -88,26 +87,39 @@ public static JavaxWebSocketServerContainer ensureContainer(ServletContext servl
         JavaxWebSocketServerContainer container = contextHandler.getBean(JavaxWebSocketServerContainer.class);
         if (container==null)
         {
-            // Find Pre-Existing (Shared?) HttpClient and/or executor
-            HttpClient httpClient = (HttpClient)servletContext.getAttribute(JavaxWebSocketServletContainerInitializer.HTTPCLIENT_ATTRIBUTE);
-            if (httpClient == null)
-                httpClient = (HttpClient)contextHandler.getServer()
-                    .getAttribute(JavaxWebSocketServletContainerInitializer.HTTPCLIENT_ATTRIBUTE);
-
-            Executor executor = httpClient == null?null:httpClient.getExecutor();
-            if (executor == null)
-                executor = (Executor)servletContext.getAttribute("org.eclipse.jetty.server.Executor");
-            if (executor == null)
-                executor = contextHandler.getServer().getThreadPool();
-
-            if (httpClient != null && httpClient.getExecutor() == null)
-                httpClient.setExecutor(executor);
+            Supplier<WebSocketCoreClient> coreClientSupplier = () ->
+            {
+                WebSocketCoreClient coreClient = contextHandler.getBean(WebSocketCoreClient.class);
+                if (coreClient == null)
+                {
+                    // Find Pre-Existing (Shared?) HttpClient and/or executor
+                    HttpClient httpClient = (HttpClient)servletContext.getAttribute(JavaxWebSocketServletContainerInitializer.HTTPCLIENT_ATTRIBUTE);
+                    if (httpClient == null)
+                        httpClient = (HttpClient)contextHandler.getServer().getAttribute(JavaxWebSocketServletContainerInitializer.HTTPCLIENT_ATTRIBUTE);
+
+                    Executor executor = httpClient == null?null:httpClient.getExecutor();
+                    if (executor == null)
+                        executor = (Executor)servletContext.getAttribute("org.eclipse.jetty.server.Executor");
+                    if (executor == null)
+                        executor = contextHandler.getServer().getThreadPool();
+
+                    if (httpClient != null && httpClient.getExecutor() == null)
+                        httpClient.setExecutor(executor);
+
+                    // create the core client
+                    coreClient = new WebSocketCoreClient(httpClient, WebSocketComponents.ensureWebSocketComponents(servletContext));
+                    coreClient.getHttpClient().setName("Javax-WebSocketClient@" + Integer.toHexString(coreClient.getHttpClient().hashCode()));
+                    if (executor != null && httpClient == null)
+                        coreClient.getHttpClient().setExecutor(executor);
+                }
```

Comment:
Should we set the coreClient as a bean/attribute here?  And/Or set the httpClient as a bean/attribute?

If we start 2 contexts with default settings and both use a websocket Javax client, then we want only a single httpClient and all its heavyweight mechanisms created.

Reply:
I would say it depends on how we want the coreClient to behave when the JavaxWebSocketServerContainer shuts down (LifeCycle stuff)

Reply:
@joakime good point.  Let's not share it by default because if shared by contexts, it is unclear who should stop it.    If we want a shared one, it can be added as a bean on the server and then set as an attribute on the contexts that we want to use it.  It will then stop with the server, not with any particular context.